Add Opower integration for getting electricity/gas usage and cost for many utilities

tronikos opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Opower integration. More than 175 utilities use Opower. Currently the following utilities are supported:

- Pacific Gas & Electric (PG&E), California's largest utility, also largest (by revenue) utility in US
- Puget Sound Energy (PSE), Washington's largest utility
- Exelon subsidiaries
  - Atlantic City Electric, New Jersey
  - Baltimore Gas and Electric (BGE), Maryland’s largest utility
  - Commonwealth Edison (ComEd), Illinois' largest utility
  - Delmarva Power, Delaware
  - PECO Energy Company (PECO), Pennsylvania's largest utility
  - Potomac Electric Power Company (Pepco), Washington, D.C. and Maryland

According to wikipedia all the above utilities serve over 15 million households.

Anyone can update the underlying library to support logging in to their utility website and this integration should automatically pick up their utility after bumping the version.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26793

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tronikos committed:
Create Opower integration

tronikos committed:
Merge branch 'dev' into opower

tronikos committed:
fix tests

Lash-L commented on homeassistant/components/opower/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+    errors = {}
+    try:
+        await api.async_login()
+    except ClientResponseError as err:
```
I think it is preferred if you handle the clientresponse error in your api and have specific api exceptions that you raise. It gives you a lot more power and is better for error catching/ makes the code in core more readable

tronikos replied:
Everybody understands HTTP status codes. I don't see much advantage of having my own exceptions that wrap them. In addition, the coordinator where you are supposed to do the polling already handles HTTP errors so I think it's fine to keep them as is.

ScottG489 replied:
A bit late to the party, but fwiw I think I'm with Lash-L on this one. It wouldn't necessarily have to be a 1:1 wrapping of the status codes. A 400 can represent many different errors for example. The lib should just expose an interface representative of it's domain and http isn't part of that.

tronikos replied:
Done.

Lash-L commented on homeassistant/components/opower/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+        login_data[CONF_USERNAME],
+        login_data[CONF_PASSWORD],
+    )
+    errors = {}
```
type errors here

Lash-L commented on homeassistant/components/opower/coordinator.py:
```diff
@@ -0,0 +1,205 @@
+            hass,
+            _LOGGER,
+            name="Opower",
+            update_interval=timedelta(hours=12),
```
How frequently does data update on opower side. 12 hours seems long.

tronikos replied:
Added comment
            # Data is updated daily on Opower.
            # Refresh every 12h to be at most 12h behind.

Lash-L commented on homeassistant/components/opower/coordinator.py:
```diff
@@ -0,0 +1,205 @@
+    ) -> dict[str, Forecast]:
+        """Fetch data from API endpoint."""
+        try:
+            await self.api.async_login()
```
Do you have to log back in each time you want data? How frequently does the login expire?

tronikos replied:
Yes. Added comment:
            # Login expires after a few minutes.
            # Given the infrequent updating (every 12h)
            # assume previous sessions have expired and re-login.

Lash-L commented on homeassistant/components/opower/coordinator.py:
```diff
@@ -0,0 +1,205 @@
+
+    async def _insert_statistics(self, accounts: list[Account]) -> None:
+        """Insert Opower statistics."""
+        if "recorder" not in self.hass.config.components:
```
I don't think this is needed? You have recorder in dependencies in the manifest

tronikos replied:
Removed. You are likely right. Tibber integration doesn't check but kitchen_sink does https://github.com/home-assistant/core/blob/dev/homeassistant/components/kitchen_sink/__init__.py#L52

Lash-L commented on homeassistant/components/opower/sensor.py:
```diff
@@ -0,0 +1,216 @@
+        name="Typical monthly gas cost",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        suggested_display_precision=0,
```
Why do you suggest 0 display precision for all of these? I would think you would at least want cost to be 2 decimal places

tronikos replied:
Added comment:
# suggested_display_precision=0 for all sensors since
# Opower provides 0 decimal points for all these.
# (for the statistics in the energy dashboard Opower does provide decimal points)

Lash-L reviewed (commented):
Just a few things I noticed, hope this helps!

Lash-L commented:
I would also add into the description why only PG&E is supported and how future utilities could be added

tronikos committed:
Update config_flow.py

tronikos committed:
Update coordinator.py

tronikos committed:
Update sensor.py

tronikos commented:
> I would also add into the description why only PG&E is supported and how future utilities could be added

Updated.

tronikos committed:
Update sensor.py

tronikos committed:
Update coordinator.py

tronikos committed:
Merge branch 'dev' into opower

tronikos committed:
Bump opower==0.0.4

borski commented:
Just following up, @Lash-L any chance we can get this reviewed? :)

tronikos committed:
Ignore errors for "recent" PGE accounts

Lash-L commented:
@borski I'm not a core maintainer. My review doesn't matter for anything other than just trying to help out

tronikos committed:
Add type for forecasts

tronikos committed:
Bump opower to 0.0.5

tronikos committed:
Bump opower to 0.0.6

tronikos committed:
Merge branch 'dev' into opower

tronikos committed:
Bump opower to 0.0.7

tronikos commented:
FYI, for anyone that will review this since https://developers.home-assistant.io/docs/review-process?_highlight=review#who-is-responsible-for-reviewing-prs recommends dropping a comment "I tested this using ... and it works", many users have started successfully using this via https://github.com/tronikos/opower-hacs See also https://community.home-assistant.io/t/pg-e-energy-monitoring-integration/336038

tronikos committed:
Merge branch 'dev' into opower

tronikos committed:
Update requirements_all.txt

tronikos committed:
Update requirements_test_all.txt

tronikos committed:
Merge branch 'dev' into opower

tronikos committed:
Update coordinator

Fix exception caused by https://github.com/home-assistant/core/pull/92095
{} is dict but the function expects a set so change it to set()

tronikos committed:
Improve exceptions handling

tronikos committed:
Merge branch 'home-assistant:dev' into opower

tronikos committed:
Bump opower==0.0.9

tronikos committed:
Bump opower to 0.0.10

tronikos committed:
Bump opower to 0.0.11

tronikos commented:
FYI, the underlying library now supports 8 utilities in US. The supported utilities serve over 15 million US households.

tronikos committed:
Merge branch 'dev' into opower

domoritz commented:
That's awesome. If I can express a wish to add support https://www.duquesnelight.com which also supports Opower. The API is at https://duq.opower.com.

tronikos commented:
Without access to an account I cannot support logging in to other utilities. You are on your own on that. Feel free to send me a PR to the underlying library https://github.com/tronikos/opower if you figure out how to programmatically login to your utility website and get the access token for opower. The current implementation for other utilities should help.

tronikos committed:
Merge branch 'dev' into opower

robbiet480 commented:
👀

balloob commented on homeassistant/components/opower/strings.json:
```diff
@@ -0,0 +1,28 @@
+      },
+      "reauth_confirm": {
+        "title": "[%key:common::config_flow::title::reauth%]",
+        "description": "Enter password for {username}",
```
We also support a read-only constant field. Will look slightly better.

tronikos replied:
I'm not familiar with that. Can you please link to an example?

balloob replied:
It's the `ConstantSelector`

balloob replied:
You can see it in action here https://design.home-assistant.io/#components/ha-form (search for Constant)

tronikos replied:
I changed the code to allow username to be changed in the reauth flow and having it prepopulated to the current value.

balloob replied:
Is there really a use case for that one? Different username but same data? And since data is synced back, should we really support that?

tronikos replied:
PG&E doesn't let you change the username. Not sure about other utilities. I changed it to constant. People can always remove and create a new entry if they can change the username.

tronikos committed:
fix issue when integration hasn't run for 30 days

use last stat time instead of now when fetching recent usage/cost

tronikos committed:
Allow username to be changed in reauth

tronikos committed:
Don't allow changing  username in reauth flow

MartinHjelmare commented on tests/components/opower/conftest.py:
```diff
@@ -0,0 +1,25 @@
+            "username": "test-username",
+            "password": "test-password",
+        },
+        state=ConfigEntryState.LOADED,
```
Don't set the config entry state here.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

MartinHjelmare commented on homeassistant/components/opower/strings.json:
```diff
@@ -0,0 +1,28 @@
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```
Please use the `already_configured_service` reference.

tronikos commented:
Thanks. Addressing your comments in https://github.com/home-assistant/core/pull/95763
